client/blockbook/client.go

```diff
@@ -22,21 +20,62 @@ import (
 	"github.com/OpenBazaar/golang-socketio/protocol"
 	"github.com/OpenBazaar/multiwallet/client/transport"
 	"github.com/OpenBazaar/multiwallet/model"
+	"github.com/btcsuite/btcd/chaincfg"
 	"github.com/btcsuite/btcd/chaincfg/chainhash"
 	"github.com/btcsuite/btcutil"
+	"github.com/cpacia/bchutil"
 	"github.com/op/go-logging"
 	"golang.org/x/net/proxy"
 )
 
 var Log = logging.MustGetLogger("client")
 
+type wsWatchdog struct {
+	client    *BlockBookClient
+	done      chan struct{}
+	wsStopped chan struct{}
+}
+
+func newWebsocketWatchdog(client *BlockBookClient) *wsWatchdog {
+	return &wsWatchdog{
+		client:    client,
+		done:      make(chan struct{}, 0),
+		wsStopped: make(chan struct{}, 0),
+	}
+}
+
+func (w *wsWatchdog) guardWebsocket() {
+	for {
+		select {
+		case <-w.wsStopped:
+			time.Sleep(1 * time.Second)
+			Log.Warningf("websocket restarting")
+			w.client.SocketClient.Close()
+			w.client.SocketClient = nil
+			w.client.setupListeners()
+			w.wsStopped = make(chan struct{}, 0)
+		case <-w.done:
+			return
+		}
+	}
+}
+
+func (w *wsWatchdog) bark() {
+	close(w.wsStopped)
```

Comment:
This will cause a panic due to `channel already closed` if two error or disconnect messages happen before the guardWebsocket loop clears it and creates a new channel.

Reply:
One quick idea: you could make `wsStopped` buffered and write to it instead of close it. Then the `guardWebsocket` loop listens for a send, drains any additional items, and restarts.